### PR TITLE
feat: add unstable Vite support

### PR DESCRIPTION
**Pull request by markdalgleish**
This PR is based on the initial Vite prototype in [pcattori/revive](https://github.com/pcattori/revive) built by @pcattori and myself.

TODO:

- [x] Finish initial docs
  - [x] Move to new Future section
  - [x] Hosting/runtime/feature compatibility matrix
  - [x] Call out HMR differences
  - [x] Thank everyone who provided prior art and/or inspired our work
- [x] Finalise API
  - [x] Rename "experimental" to "unstable"
  - [x] Move plugin from `@remix-run/dev/vite` to `@remix-run/dev`
  - [x] Remove Remark plugin (for now at least)
- [x] Dedupe config resolution logic being used inside Vite plugin
- [x] Warn about unstable usage in terminal

### CODE REVIEW

**Comment by OnurGvnc** [comment]
[hono](https://hono.dev/) has a vite-dev-server that is compatible with cloudflare workers. I don't know all the details, but here are some links that might give an idea about cloudflare workers compatibility
- https://www.npmjs.com/package/@hono/vite-dev-server
- https://github.com/honojs/vite-plugins/tree/main/packages/dev-server
- https://github.com/honojs/vite-plugins/blob/f5d4c78cfc94b4e46f9dd191f1cd07da832cdee5/packages/dev-server/src/dev-server.ts#L46-L53
- https://github.com/yusukebe/vite-dev-server-examples/blob/main/src/index.ts

maybe related:
- https://github.com/cloudflare/miniflare/pull/639
- https://github.com/leader22/cfw-bindings-wrangler-bridge
- https://github.com/james-elicx/cf-bindings-proxy

File: docs/future/vite.md [hunk 1]
```diff
@@ -0,0 +1,522 @@
+---
+title: Vite (Unstable)
+toc: false
+---
+
+# Vite (Unstable)
+
+<docs-warning>Vite support is currently unstable and only intended to gather early feedback. We don't yet recommend using this in production.</docs-warning>
+
+[Vite] is a powerful, performant and extensible development environment for JavaScript projects. In order to improve and extend Remix's bundling capabilities, we're currently exploring the use of Vite as an alternative compiler to esbuild.
```
Comment (pcattori) [inline docs/future/vite.md:10]:
```suggestion
[Vite][vite] is a powerful, performant and extensible development environment for JavaScript projects. In order to improve and extend Remix's bundling capabilities, we're currently exploring the use of Vite as an alternative compiler to esbuild.
```

**Review by pcattori — CHANGES_REQUESTED** [review]

File: docs/future/vite.md [hunk 2]
```diff
@@ -0,0 +1,522 @@
...
+  ...(cssBundleHref
+    ? [{ rel: "stylesheet", href: cssBundleHref }]
+    : []),
+  // ...
+];
+```
+
+### Tailwind
+
+To use [Tailwind] in Vite, first install the required dependencies:
```
Comment (pcattori) [inline docs/future/vite.md:189]: Missing link:

```suggestion
To use [Tailwind][tailwind] in Vite, first install the required dependencies:
```

**Review by pcattori — CHANGES_REQUESTED** [review]

File: docs/future/vite.md [hunk 3]
```diff
@@ -0,0 +1,522 @@
...
+      name: "description",
+      content: frontmatter.description,
+    },
+  ];
+};
+
+# Hello World
+```
+
+<docs-info>You may even want to take this a step further and create a [remark plugin][remark-plugin] that automatically maps frontmatter to route exports for you.</doc-info>
```
Comment (pcattori) [inline docs/future/vite.md:360]:
```suggestion
```

**Review by pcattori — CHANGES_REQUESTED** [review]

File: docs/future/vite.md [hunk 4]
```diff
@@ -0,0 +1,519 @@
...
+const posts = import.meta.glob("./posts/*.mdx", {
+  eager: true,
+});
+```
+
+### `*.server.ts` and `*.client.ts` Extensions
+
+The Remix compiler allowed you to define server/client-only files using the `*.server.ts`/`*.client.ts` extensions. The Remix Vite plugin supports this pattern with one notable difference.
+
+Since Vite leverages ESM at runtime to load modules, you may need to use `import *` when importing server/client-only files if the import statement hasn't been removed by dead code elimination.
```
Comment (pcattori) [inline docs/future/vite.md:393]:
```suggestion
```

**Review by pcattori — CHANGES_REQUESTED** [review]

File: docs/future/vite.md [hunk 5]
```diff
@@ -0,0 +1,519 @@
...
+// these exports are specially handled by Remix to be HMR-compatible
+export const meta = { title: "Home" }; // ✅
+export const links = [
+  { rel: "stylesheet", href: "style.css" },
+]; // ✅
+export const headers = { "Cache-Control": "max-age=3600" }; // ✅
+
+// these exports are treeshaken by Remix, so they never affect HMR
+export const loader = () => {}; // ✅
+export const action = () => {}; // ✅
```
Comment (pcattori) [inline docs/future/vite.md:441]:
```suggestion

// these exports are treeshaken by Remix, so they never affect HMR
export const headers = { "Cache-Control": "max-age=3600" }; // ✅
export const loader = () => {}; // ✅
export const action = () => {}; // ✅
```

**Review by pcattori — CHANGES_REQUESTED** [review]

**Review by pcattori — APPROVED** [review]

File: docs/future/vite.md [hunk 6]
```diff
@@ -0,0 +1,517 @@
...
+export const meta = { title: "Home" }; // ✅
+export const links = [
+  { rel: "stylesheet", href: "style.css" },
+]; // ✅
+export const headers = { "Cache-Control": "max-age=3600" }; // ✅
+
+// These exports are removed by the Remix Vite plugin
+// so they never affect HMR
+export const loader = () => {}; // ✅
+export const action = () => {}; // ✅
```
Comment (pcattori) [inline docs/future/vite.md:440]: Headers is server-only:

```suggestion
// These exports are handled by the Remix Vite plugin
// to be HMR-compatible
export const meta = { title: "Home" }; // ✅
export const links = [
  { rel: "stylesheet", href: "style.css" },
]; // ✅

// These exports are removed by the Remix Vite plugin
// so they never affect HMR
export const headers = { "Cache-Control": "max-age=3600" }; // ✅
export const loader = () => {}; // ✅
export const action = () => {}; // ✅
```

**Review by pcattori — CHANGES_REQUESTED** [review]

File: docs/future/vite.md (hunk 5 above)
Comment (pcattori) [reply docs/future/vite.md]: outdated. added a [separate suggestion here](https://github.com/remix-run/remix/pull/7590#discussion_r1357743458)

**Review by pcattori — APPROVED** [review]

**Comment by mmachatschek** [comment]
@pcattori where should issues be reported?

**Comment by pcattori** [comment]
Yea separate issue with "Vite Nightly" or something like that is the way to go

**Pull request merged**